### ISSUE TITLE
\str_set_convert:Nnnn produce invalid str variable if there is a space

### ISSUE THREAD

**Issue by u-fischer** [latex3/latex3#1344]
~~~~
\documentclass{article}
\begin{document}
\ExplSyntaxOn
\str_set_convert:Nnnn\l_tmpb_str{hallo~world}{utf8}{default/bytes}
\str_show:N\l_tmpb_str
\ExplSyntaxOff
\end{document}
~~~~

errors with 

~~~~
! LaTeX Error: Variable '\l_tmpb_str' is not a valid str.
~~~~

(I'm not sure if this a problem with the conversion or with the `\str_show:N`.)

**Comment by muzimuzhi** [comment]
Space character in conversion result is mistakenly set to catcode 12 (other).
```tex
\documentclass{article}
\begin{document}
\ExplSyntaxOn
\str_set_convert:Nnnn\l_tmpb_str{hallo~world}{utf8}{default/bytes}
\tl_analysis_show:N \l_tmpb_str
\ExplSyntaxOff
\end{document}

The token list \l_tmpb_str contains the tokens:
>  h (the character h)
>  a (the character a)
>  l (the character l)
>  l (the character l)
>  o (the character o)
>    (the character  ) % <<< wrong catcode
>  w (the character w)
>  o (the character o)
>  r (the character r)
>  l (the character l)
>  d (the character d).
<recently read> }
```

Probably caused by how `\__str_encode_native_char:n` is defined
-  In XeTeX or LuaTeX:
  https://github.com/latex3/latex3/blob/3bf9a215856498a97372c67838cea926f14ffe30/l3kernel/l3str-convert.dtx#L992-L993
- Otherwise
  https://github.com/latex3/latex3/blob/3bf9a215856498a97372c67838cea926f14ffe30/l3kernel/l3str-convert.dtx#L1003-L1011

**Comment by josephwright** [comment]
Use `\codepoint_str_generate:n`?

**Comment by josephwright** [comment]
Or perhaps as that is relatively slow just put a wrapper in that sets catcode as required?

**Comment by u-fischer** [comment]
> Space character in conversion result is mistakenly set to catcode 12 (other).

Yes, I saw that. The question is if it is "mistakenly" or by design (e.g. for speed), and if the \show command should error on a catcode 12 space in a string?

**Comment by josephwright** [comment]
@u-fischer  I think the code is in error here: we are clear that a `str` means catcode-10 spaces, catcode-12 otherwise, cf. `\c_catcode_other_space_tl`.

**Comment by muzimuzhi** [comment]
`\c__str_byte_32_tl` is also mistakenly defined.
https://github.com/latex3/latex3/blob/3bf9a215856498a97372c67838cea926f14ffe30/l3kernel/l3str-convert.dtx#L463-L481

Something like
```tex
\cs_undefine:c { c__str_byte_32_tl }
\tl_const:cn { c__str_byte_32_tl } { { ~ } 2 0 }
```
seems to do the trick.

```tex
\documentclass{article}

\ExplSyntaxOn
% patches to \__str_encode_native_char:n for unicode-aware engines (?)

% patch to \c__str_byte_32_tl
\cs_undefine:c { c__str_byte_32_tl }
\tl_const:cn { c__str_byte_32_tl } { { ~ } 2 0 }

% test
\str_set_convert:Nnnn \l_tmpb_str {hallo~world} {utf8} {default}
\str_show:N \l_tmpb_str
\ExplSyntaxOff

\begin{document}
\end{document}
```

Still not sure if all cases are covered. Sorry I don't know enough about encodings to be confident enough to prepare a PR.

**Comment by josephwright** [comment]
Basic ideas look OK, but the test suite is not happy: we end up loosing spaces. I'm working on it.

**Comment by josephwright** [comment]
Lots of places assume no spaces ...

**Comment by josephwright** [comment]
I wonder if a better fix is to apply `\tl_to_str:n` at the end of the process to re-normalise spaces. This avoids having to re-work significant parts of the internal loops: thoughts?

**Comment by muzimuzhi** [comment]
Sounds a nice trick.

**Comment by josephwright** [comment]
Fixed by 5ee772c4e7bdfdc1d2d32a6ffaaa8d0a513d8bff
